src/main/java/net/minestom/server/entity/Player.java

```diff
@@ -2501,29 +2521,43 @@ public long getLastKeepAlive() {
     }
 
     /**
-     * Gets the packet to add the player from the tab-list.
+     * Gets the packet to render all the players that are online.
      *
-     * @return a {@link PlayerInfoPacket} to add the player
+     * @return a {@link PlayerInfoPacket} containing all online players
      */
-    @NotNull
-    protected PlayerInfoPacket getAddPlayerToList() {
+    protected @NotNull PlayerInfoPacket getAddAllPlayerToList() {
         PlayerInfoPacket playerInfoPacket = new PlayerInfoPacket(PlayerInfoPacket.Action.ADD_PLAYER);
+        for (Player player : MinecraftServer.getConnectionManager().getOnlinePlayers()) {
```

Comment:
Same as before, all players loop are terrible

Reply:
In order for player skins to display, we currently add all players to a user's tab when they join then remove them if necessary.

I don't know of a better way to do this unless there's some way to tell when a user comes into another user's render distance server side.

Reply:
`Player#addViewer`?

Reply:
So in the addViewer method we then do the add/remove if they're not on the tablist? Then additionally this means we'll end up doing this multiple times unless we keep a cache of every player skin each player has received.

Reply:
Why cannot you trust the tablist populator? Feels like it should manage which players you need to send instead of the whole server.
I am probably missing something

Reply:
Yh. It's designed so you don't have to add everyone that a player can see to the tablist. E.g you could only display your faction members in the tablist.

We still have to send all player skins as otherwise, say they go to spawn and see other users, they would all have alex/steve skins.

Reply:
I don't think that sending a new PlayerInfo packet in `Player#addViewer` is an issue, caching should be unnecessary.

So as you said, just sending the players you need to see in the tablist on login, and then ensure that player can be seen in the addViewer method

Reply:
Sounds good 👍🏻. @joeyoey if you've got time for this otherwise I'll do it in a few days.

Reply:
Will do, basically there can be multiple tablists and each tab list can have as many or as few players as are chosen. However, typically without zak and I’s api a player that is not on the tab list can not be seen as the client sees the other player as invisible or Steve/Alex since they’re not on their tab list, so to avoid this we HAVE to cache to keep the skins and other players visible. However the populator is over complicated for what it is, so I’ll move the add/remove of players from tablists to the addviewer, removeviewer, player#init, and player#disconnect/player#remove. Not sure about the last one as in preliminary tests the removal of the player on disconnect breaks some things. I’ll debug and figure it out in the next few days.

Reply:
So, is the all-player loop still necessary? I expect it to be a huge bottleneck with high amount of players

Reply:
Does Player#addViewer get called when the player logs in? If so I can remove it entirely.